src/crypto/openssl/cose_signer.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+
+#include "crypto/openssl/cose_signer.h"
+
+#include "ccf/ds/logger.h"
+
+#include <openssl/evp.h>
+#include <t_cose/t_cose_sign1_sign.h>
+
+namespace ccf::crypto
+{
+  using namespace OpenSSL;
+
+  COSESigner_OpenSSL::COSESigner_OpenSSL(const Pem& key) :
+    key_pair(std::make_shared<KeyPair_OpenSSL>(key))
+  {}
+
+  bool COSESigner_OpenSSL::sign(
+    std::span<uint8_t> payload, std::vector<uint8_t>& cose_envelope) const
+  {
+    struct t_cose_sign1_sign_ctx sign_ctx;
+    t_cose_sign1_sign_init(&sign_ctx, 0, T_COSE_ALGORITHM_ES256);
+
+    EVP_PKEY* evp_key = *key_pair;
+
+    t_cose_key signing_key;
+    signing_key.crypto_lib = T_COSE_CRYPTO_LIB_OPENSSL;
+    signing_key.k.key_ptr = evp_key;
+
+    t_cose_sign1_set_signing_key(&sign_ctx, signing_key, NULL_Q_USEFUL_BUF_C);
+
+    // Neither documentation or code example explain how much memory should be
+    // reserved. Encoding will fail gracefully if the buffer was too small.
+    const size_t payload_buffer_size{payload.size() * 2 + 128};
+
+    Q_USEFUL_BUF_MAKE_STACK_UB(constructed_payload_buffer, payload_buffer_size);
+    q_useful_buf_c constructed_payload;
+
+    q_useful_buf_c payload_to_encode{payload.data(), payload.size()};
+
+    QCBOREncodeContext cbor_encode;
+    QCBOREncode_Init(&cbor_encode, constructed_payload_buffer);
+    QCBOREncode_OpenMap(&cbor_encode);
+    QCBOREncode_AddBytes(&cbor_encode, payload_to_encode);
+    QCBOREncode_CloseMap(&cbor_encode);
+
+    auto qcbor_error = QCBOREncode_Finish(&cbor_encode, &constructed_payload);
+    if (qcbor_error != QCBOR_SUCCESS)
+    {
+      LOG_DEBUG_FMT(
+        "QCBOR payload encoding failed with error code {}", qcbor_error);
+      return false;
+    }
+
+    // https://laurencelundblade.github.io/t_cose/t__cose__sign1__sign_8h.html
+    // says the total signature overhead is `about` 150 bytes, so considering
+    // 300 as a somewhat safe value here.
+    const size_t cose_buffer_size{payload_buffer_size + 300};
+    Q_USEFUL_BUF_MAKE_STACK_UB(signed_cose_buffer, cose_buffer_size);
+    struct q_useful_buf_c signed_cose;
+
+    auto cose_error = t_cose_sign1_sign(
+      &sign_ctx, constructed_payload, signed_cose_buffer, &signed_cose);
+
+    if (cose_error != T_COSE_SUCCESS)
+    {
+      LOG_DEBUG_FMT("COSE Sign1 failed with error code {}", cose_error);
+      return false;
+    }
+
+    cose_envelope.resize(signed_cose.len);
```

Comment:
nit: combining the two: cose_envelope.assign(signed_cose.ptr, signed_cose.ptr + signed_cose.len);

Reply:
This got changed